zilliqa/src/consensus.rs

```diff
@@ -999,83 +1032,148 @@ impl Consensus {
+        // Retrieve parent block data
+        let parent_block = self
+            .get_block(&proposal.parent_hash())?
+            .context("missing parent block")?;
+        let parent_block_hash = parent_block.hash();
+        let parent_block_view = parent_block.view();
+
+        // Should have supermajority by now.
+        // Retrieve the committee - for rewards
+        let committee = state.get_stakers_at_block_raw(&parent_block)?;
+        let (signatures, cosigned, _, _) = votes
+            .get(&parent_block_hash)
+            .context("tried to finalise a proposal without any votes")?;
+
+        // Compute the majority QC
+        let qc = self.qc_from_bits(parent_block_hash, signatures, *cosigned, parent_block_view);
+
+        // Retrieve the previous leader - for rewards.
+        let proposer = self
+            .leader_at_block(&parent_block, parent_block_view + 1)
+            .context("missing parent block leader")?
+            .public_key;
 
-        let block_hash = block.hash();
-        let block_view = block.view();
+        // Apply the rewards when exiting the round
+        let previous_state_root_hash = state.root_hash()?;
+        state.set_to_root(proposal.state_root_hash().into()); // proposal late state to apply rewards
+        Self::apply_rewards_late_at(
+            parent_block.state_root_hash(),
+            state,
+            &self.config.consensus,
+            &committee,
+            proposer.into(), // Last leader
+            parent_block_view + 1,
+            &qc.cosigned, // QC cosigners
+        )?;
 
-        let committee: Vec<_> = state.get_stakers_at_block_raw(&block)?;
+        // Finalise the proposal with final QC and state.
+        let proposal = Block::from_qc(
+            self.secret_key,
+            proposal.header.view,
+            proposal.header.number,
+            // majority QC
+            qc,
+            // post-reward updated state
+            state.root_hash()?,
+            proposal.header.transactions_root_hash,
+            proposal.header.receipts_root_hash,
+            proposal.transactions,
+            proposal.header.timestamp,
+            proposal.header.gas_used,
+            proposal.header.gas_limit,
+        );
 
-        let Some((signatures, cosigned, cosigned_weight, supermajority_reached)) =
-            votes.get(&block_hash).cloned()
-        else {
-            warn!(%block_hash, %block_view, "tried to create a proposal without any votes, this shouldn't happen");
-            return Ok(None);
-        };
+        state.set_to_root(previous_state_root_hash.into());
 
-        let qc = self.qc_from_bits(block_hash, &signatures, cosigned, block_view);
-        let parent_hash = qc.block_hash;
+        // Return the final proposal
+        Ok(Some(proposal))
+    }
+
+    /// Assembles the Proposal block early.
+    /// This is performed before the majority QC is available.
+    /// It does all the needed work but with a dummy QC.
```

Comment:
What does dummy QC mean? A QC which doesn't represent the supermajority but is otherwise still valid?